compiler/util/canonicalize_affine.py

```diff
@@ -0,0 +1,121 @@
+            # both are constants: fold constants
+            return AffineConstantExpr(expr.lhs.value + expr.rhs.value)
+        else:
+            # move constant to rhs
+            expr = AffineBinaryOpExpr(expr.kind, expr.rhs, expr.lhs)
+    if isinstance(expr.rhs, AffineConstantExpr):
+        # rhs is constant, lhs is not
+        # addition by 0 can be omitted
+        if expr.rhs.value == 0:
+            return expr.lhs
+    # order by minimum dimension first
+    # d0 + d1: good. d1 + d0: bad
+    dim_lhs = get_dim(expr.lhs)
+    dim_rhs = get_dim(expr.rhs)
+    if dim_rhs is not None:
+        if dim_lhs is None or dim_lhs > dim_rhs:
+            return expr.rhs + expr.lhs
+    return expr
+
+
+def canonicalize_multiplication(expr: AffineBinaryOpExpr) -> AffineExpr:
+    # always put the constant on rhs
+    assert expr.kind is AffineBinaryOpKind.Mul
+    if isinstance(expr.lhs, AffineConstantExpr):
+        # lhs is constant
+        if isinstance(expr.rhs, AffineConstantExpr):
+            # both are constants: fold constants
+            return AffineConstantExpr(expr.lhs.value * expr.rhs.value)
+        else:
+            # move constant to rhs
+            return AffineBinaryOpExpr(expr.kind, expr.rhs, expr.lhs)
+    if isinstance(expr.rhs, AffineConstantExpr):
+        # rhs is constant, lhs is not
+        # multiplication by 1 can be omitted
+        if expr.rhs.value == 1:
+            return expr.lhs
+    return expr
+
+
+def canonicalize_floordiv(expr: AffineBinaryOpExpr) -> AffineExpr:
+    assert expr.kind is AffineBinaryOpKind.FloorDiv
+    if isinstance(expr.rhs, AffineConstantExpr):
+        # division by 1 can be omitted
+        if expr.rhs.value == 1:
+            return expr.lhs
+    return expr
+
+
+def canonicalize_mod(expr: AffineBinaryOpExpr) -> AffineExpr:
+    assert expr.kind is AffineBinaryOpKind.Mod
+    if isinstance(expr.rhs, AffineConstantExpr):
+        # module 1 is always 0
+        if expr.rhs.value == 1:
+            return AffineConstantExpr(0)
+    return expr
+
+
+def canonicalize_binary_op(expr: AffineBinaryOpExpr) -> AffineExpr:
+    # canonicalize children
+    expr = AffineBinaryOpExpr(
+        expr.kind, canonicalize_expr(expr.lhs), canonicalize_expr(expr.rhs)
+    )
+    if expr.kind is AffineBinaryOpKind.Add:
+        return canonicalize_addition(expr)
+    if expr.kind is AffineBinaryOpKind.Mul:
+        return canonicalize_multiplication(expr)
+    if expr.kind is AffineBinaryOpKind.FloorDiv:
+        return canonicalize_floordiv(expr)
+    if expr.kind is AffineBinaryOpKind.Mod:
+        return canonicalize_mod(expr)
+    return expr
+
+
+def canonicalize_expr(expr: AffineExpr) -> AffineExpr:
+    if isinstance(expr, AffineBinaryOpExpr):
+        return canonicalize_binary_op(expr)
+
+    return expr
+
+
+# helper function to canonicalize affine maps
+def canonicalize_map(map: AffineMap) -> AffineMap:
+    # canonicalize each result and construct new affine map
```

Comment:
I'm not sure what this means :(

Reply:
An affine map is built out of multiple affine expressions. These are called `results`. So we canonicalize an affine map by canonicalizing all o the expressions / results.